flocker/node/agents/cinder.py

```diff
@@ -457,37 +487,140 @@ def allocation_unit(self):
+                    u"compute_instance_id:metadata_response_error"),
+                status_code=response.status_code,
+                content=response.content,
+            ).write()
+
+    def _compute_instance_id_from_ipaddresses(self):
         """
-        Find the ``ACTIVE`` Nova API server with a subset of the IPv4 and IPv6
-        addresses on this node.
+        Find the ``ACTIVE`` Nova API server with an intersection of
+        the IPv4 and IPv6 addresses on this node.
         """
         local_ips = get_all_ips()
         api_ip_map = {}
-        matching_instances = []
+        id_to_node_ips = {}
         for server in self.nova_server_manager.list():
             # Servers which are not active will not have any IP addresses
             if server.status != u'ACTIVE':
                 continue
             api_addresses = _extract_nova_server_addresses(server.addresses)
-            # Only do subset comparison if there were *some* IP addresses;
-            # non-ACTIVE servers will have an empty list of IP addresses and
-            # lead to incorrect matches.
-            if api_addresses and api_addresses.issubset(local_ips):
-                matching_instances.append(server.id)
-            else:
-                for ip in api_addresses:
-                    api_ip_map[ip] = server.id
-
-        # If we've got this correct there should only be one matching instance.
-        # But we don't currently test this directly. See FLOC-2281.
-        if len(matching_instances) == 1 and matching_instances[0]:
-            return matching_instances[0]
-        # If there was no match, or if multiple matches were found, log an
-        # error containing all the local and remote IPs.
-        COMPUTE_INSTANCE_ID_NOT_FOUND(
-            local_ips=local_ips, api_ips=api_ip_map
-        ).write()
+            id_to_node_ips[server.id] = api_addresses
+            for ip in api_addresses:
+                api_ip_map[ip] = server.id
+
+        try:
+            return _get_compute_id(local_ips, id_to_node_ips)
+        except KeyError:
+            # If there was no match, or if multiple matches were found, log an
+            # error containing all the local and remote IPs.
+            COMPUTE_INSTANCE_ID_NOT_FOUND(
+                local_ips=local_ips, api_ips=api_ip_map
+            ).write()
+
+    def compute_instance_id(self):
+        """
+        Attempt to find the local instance ID:
+        * from local configdrive if it exists, or
+        * from metadata service, or
+        * by matching the local IP addresses to those reported by the Nova API.
+        """
+        methods = [
+            self._compute_instance_id_from_configdrive,
+            self._compute_instance_id_from_metadata_service,
+            self._compute_instance_id_from_ipaddresses,
+        ]
+        for method in methods:
+            result = method()
+            if result is not None:
+                return result
```

Comment:
if configdrive or metadata service aren't available, then still theres a concern of overlapping IPs with intersection when using overlays correct? Looking at the code again it seems its goes through server by server and therefore may choose the wrong server if it is the first one with the intersection.

Maybe worth a comment or a seperate issue but I dont think its worth blocking this PR over. Recommendation would be for metadata or configdrive etc as you mention in the PR comments.